### PR TITLE
feat(web): add the document-editor layout model

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title (required, present tense):

  ```text
  feat(web): add the document-editor layout model
  ```

- Type:
  - [x] feat (minor)
  - [ ] fix / perf (patch)
  - [ ] docs
  - [ ] refactor
  - [ ] test
  - [ ] chore / ci / style

- Breaking change:
  - [ ] `!` in title or `BREAKING CHANGE:` footer included

## What's Changing

Adds `apps/web/src/lib/docLayout.ts`, a pure, store-agnostic module that turns a
`ResumeData` plus a template's layout metadata into the concrete page/column
structure the document editor renders — the shared layout truth every later
slice of the document-editor epic builds on.

Every export is pure and synchronous: no store reads, no Solid reactivity, no
DOM, no fetching. The module owns no mutations; it computes what the next
`metadata.layout` *should* be and the caller hands that to
`resumeStore.updateLayout`.

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [ ] Docs updated if user-facing
- [x] Local CI passed (`make test` and `uv run lintro chk`)

## Closes

- Closes #726

## Details

Export contract:

- `FIXED_SECTION_IDS` / `SECTION_LABELS` — canonical fixed-section ids and
  labels, derived from `FIXED_LAYOUT_SECTION_KEYS` in `stores/resume` so there
  is no second, divergent list.
- `layoutPages` — pages → columns → section ids, seeded from `metadata.layout`
  and falling back to the template's `defaultColumns` when it is absent or
  empty. Expands the `custom` sentinel that templates emit, drops ids that
  address no section, and keeps only the first occurrence of a repeated id.
- `renderPages` — `layoutPages` minus hidden sections, minus sections with no
  content, minus pages left entirely empty. Empty columns survive so column
  indices stay meaningful.
- `layoutColumns` — column count, relative widths and paint order from
  `layoutMode` / `sidebarWidth`. A fixed sidebar width is converted against the
  same nominal A4 content width used by `templates/_common.typ` and
  `ThemeEditor.tsx`; proportional templates use the 1/3 default.
- `layoutForTemplate` — a fresh `metadata.layout` for a template switch that
  re-places any custom section the template defaults never mention, so none is
  silently dropped.
- `findSectionPlacement`, `sectionVisible`, `sectionTitle`, `isCustomId`,
  `nextId`, `emptyItemFor` — custom sections included throughout.

`TemplateLayout` mirrors the shape now served by `GET /api/templates`
(`LayoutInfo` in `crates/server/src/dto.rs`, landed in #737) and is always
injected as a parameter. Section-level placement stays the only pagination
mechanism — no item-level page breaks.

### Testing strategy

`apps/web/src/lib/__tests__/docLayout.test.ts` adds 50 Vitest cases, one block
per export, over the shared corpus `tests/fixtures/v3/doc-editor.json`. That
fixture is stored in the Reactive Resume v3 wire format, which Vitest cannot run
the Rust parser over, so the test carries a small adapter that mirrors the
parser (summary moves out of `basics`; custom items map `title`/`subtitle` onto
`name`/`description`).

Verified locally: `uv run lintro chk` clean, `cargo test --workspace` green, and
`bun run test` (67 files, 649 tests), `bun run lint`, `bun run typecheck` all
green in `apps/web`.

## Summary by CodeRabbit

* **New Features**
  * Added document layout support for single-column, sidebar, proportional, and header-split templates.
  * Added automatic pagination, column sizing, section ordering, placement, visibility, and titles.
  * Added support for custom sections, generated identifiers, template switching, and blank section items.

* **Bug Fixes**
  * Improved handling of duplicate or unknown sections and removed empty or invalid content from layouts.

* **Tests**
  * Added comprehensive automated coverage for document layout behavior and edge cases.

### CODE REVIEW

File: apps/web/src/lib/__tests__/docLayout.test.ts [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+
+    expect(layoutPages(resume, SIDEBAR_TEMPLATE)).toEqual([[["experience", "skills"], []]]);
+  });
+
+  it("drops ids that address no section", () => {
+    const resume = loadFixture();
+    resume.metadata.layout = [[["experience", "not-a-section"]]];
+
+    expect(layoutPages(resume, SIDEBAR_TEMPLATE)).toEqual([[["experience"]]]);
+  });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/__tests__/docLayout.test.ts:336]: Fixed in 11aa323 — added both cases: a stored layout of empty columns now pins the `hasAnySection` fallback, and a layout listing the sentinel in two columns pins single expansion (`speaking`, `advisory`, in fixture key order).

File: apps/web/src/lib/__tests__/docLayout.test.ts [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+    expect(pages[1][1]).toEqual(["languages", "interests", "certifications"]);
+  });
+
+  it("drops sections with no content", () => {
+    const resume = loadFixture();
+    resume.sections.awards.items = [];
+    resume.metadata.layout = [[["awards", "experience"]]];
+
+    expect(renderPages(resume, SIDEBAR_TEMPLATE)).toEqual([[["experience"]]]);
+  });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/__tests__/docLayout.test.ts:355]: Fixed in 11aa323 — added a two-column case where the sidebar is emptied by filtering and the result keeps `[]` at index 1, so the contract `layoutColumns` depends on is now pinned.

File: apps/web/src/lib/__tests__/docLayout.test.ts [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+
+  it("shares width evenly when a page carries more columns than the mode implies", () => {
+    const columns = layoutColumns([["a"], ["b"], ["c"]], SIDEBAR_TEMPLATE);
+
+    expect(columns).toHaveLength(3);
+    for (const column of columns) {
+      expect(column.width).toBeCloseTo(1 / 3, 10);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/__tests__/docLayout.test.ts:461]: Fixed in 11aa323 — the expectation is derived (`180 / (595.28 - 2 * 18)` at 10 digits) and both clamp bounds now have cases.

File: apps/web/src/lib/__tests__/docLayout.test.ts [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+
+  it("does not mutate the resume", () => {
+    const resume = loadFixture();
+    const before = JSON.stringify(resume.metadata.layout);
+
+    layoutForTemplate(resume, SINGLE_TEMPLATE);
+
+    expect(JSON.stringify(resume.metadata.layout)).toBe(before);
+  });
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/__tests__/docLayout.test.ts]: Fixed in 11aa323 — the test now also asserts `SINGLE_TEMPLATE.defaultColumns` is unchanged.

File: apps/web/src/lib/docLayout.ts [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+ * only computes what the next `metadata.layout` *should* be — the write goes
+ * through `resumeStore.updateLayout`).
+ *
+ * `metadata.layout` is `string[][][]`: pages -> columns -> section ids. Section
+ * placement is the only pagination mechanism; there are no item-level page
+ * breaks.
+ */
+
+import { FIXED_LAYOUT_SECTION_KEYS, isHtmlEmpty } from "../stores/resume";
+import { createEmptyUrl } from "../wasm/types";
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/docLayout.ts]: Fixed in 11aa323 — `FIXED_LAYOUT_SECTION_KEYS` and `isHtmlEmpty` now live in `apps/web/src/lib/resumeSections.ts`. The store, `docLayout`, `coverLetterExport`, and the store test import from there, so nothing re-exports them and this module no longer pulls in the Solid store, the wasm bindings, or the UI toaster.

File: apps/web/src/lib/docLayout.ts [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+  if (missing.length > 0) {
+    if (columns.length === 0) {
+      columns.push([...missing]);
+    } else {
+      columns[0] = [...columns[0], ...missing];
+    }
+  }
+
+  return [columns];
+}
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply apps/web/src/lib/docLayout.ts:416]: Valid, and confirmed against the server: `coverLetter` appears in no template's `default_columns` (`crates/render/src/typst_engine/template_layout.rs` builds them from `MAIN_SECTIONS`/`SIDEBAR_SECTIONS`, neither of which lists it), and `has-cover-letter` in `templates/_common.typ` requires the id to be present in the layout, so the cover-letter page really did stop rendering after a switch. Fixed in 11aa323 along the lines you proposed, plus a regression test. Collapsing to a single page is intentional — the new template's page-0 arrangement is the point of the call — so I documented that on the function rather than changing it.
